lighthouse-core/audits/seo/canonical.js

```diff
@@ -0,0 +1,167 @@
+      helpText: 'Canonical links suggest which URL to show in search results. ' +
+        'Read more in [Use canonical URLs]' +
+        '(https://support.google.com/webmasters/answer/139066).',
+      requiredArtifacts: ['Canonical', 'Hreflang'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+
+    return artifacts.requestMainResource(devtoolsLogs)
+      .then(mainResource => {
+        const baseURL = new URL(mainResource.url);
+        let canonicals = [];
+        let hreflangs = [];
+
+        mainResource.responseHeaders
+          .filter(h => h.name.toLowerCase() === LINK_HEADER)
+          .forEach(h => {
+            canonicals = canonicals.concat(getCanonicalLinksFromHeader(h.value));
+            hreflangs = hreflangs.concat(getHreflangsFromHeader(h.value));
+          });
+
+        canonicals = canonicals.concat(artifacts.Canonical);
+
+        artifacts.Hreflang.forEach(({href}) => hreflangs.push(href));
+
+        hreflangs = hreflangs
+          .filter(href => isValidRelativeOrAbsoluteURL(href))
+          .map(href => (new URL(href, baseURL)).href); // normalize URLs
+
+        if (canonicals.length === 0) {
+          return {
+            rawValue: true,
+            notApplicable: true,
+          };
+        }
+
+        if (canonicals.length > 1) {
+          return {
+            rawValue: false,
+            displayValue: `multiple URLs (${canonicals.join(', ')})`,
+          };
+        }
+
+        const canonical = canonicals[0];
+
+        if (!isValidRelativeOrAbsoluteURL(canonical)) {
+          return {
+            rawValue: false,
+            displayValue: `invalid URL (${canonical})`,
+          };
+        }
+
+        if (!URL.isValid(canonical)) {
+          return {
+            rawValue: false,
+            displayValue: `relative URL (${canonical})`,
+          };
+        }
+
+        const canonicalURL = new URL(canonical);
+
+        // cross-language or cross-country canonicals are a common issue
+        if (hreflangs.includes(baseURL.href) && hreflangs.includes(canonicalURL.href) &&
+          baseURL.href !== canonicalURL.href) {
+          return {
+            rawValue: false,
+            displayValue: `points to another hreflang location (${baseURL.href})`,
+          };
+        }
+
+        // bing and yahoo don't allow canonical URLs pointing to different domains, it's also
+        // a common mistake to publish a page with canonical pointing to e.g. a test domain or localhost
+        if (getPrimaryDomain(canonicalURL) !== getPrimaryDomain(baseURL)) {
+          return {
+            rawValue: false,
+            displayValue: `points to a different domain (${canonicalURL})`,
+          };
+        }
+
+        // another common mistake is to have canonical pointing from all pages of the website to its root
```

Comment:
still somewhat 😢  that this can fail for intentional usage. though I don't have a real life example so it's probably fine :)